enhancements/local-storage/subprovisioner-integration-into-lvms.md

```diff
@@ -0,0 +1,255 @@
+
+
+### Workflow of Subprovisioner instantiation via LVMCluster
+
+1. The user is informed of the intended use case of Subprovisioner, and decides to use it for its multi-node capabilities before provisioning Storage
+2. The user configures LVMCluster with non-default values for the Volume Group and the deviceClass policy field
+3. The lvm-operator detects the new deviceClass policy field and creates a new CSI driver deployment for Subprovisioner.
+4. The Subprovisioner CSI driver is configured to use the shared storage deviceClass, initializes the global lock space, and starts provisioning shared storage.
+5. The user can now provision shared storage via Subprovisioner on the OpenShift Container Platform cluster.
+6. The user can also provision regular TopoLVM deviceClasses side-by-side with shared storage deviceClasses in the same cluster. Then, TopoLVM gets provisioned side-by-side.
+
+## Design Details for `LVMCluster CR extension`
+
+API scheme for `LVMCluster` CR:
+
+```go
+
+  // The DeviceAccessPolicy type defines the accessibility of the create lvm2 volume group backing the deviceClass. 
+  type DeviceAccessPolicy string
+  
+  const (
+    DeviceAccessPolicyShared DeviceAccessPolicy = "shared"
+    DeviceAccessPolicyNodeLocal  DeviceAccessPolicy = "nodeLocal"
+  )
+
+  // LVMClusterSpec defines the desired state of LVMCluster
+  type LVMClusterSpec struct {
+    // Important: Run "make" to regenerate code after modifying this file
+    
+    // Tolerations to apply to nodes to act on
+    // +optional
+    Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+    // Storage describes the deviceClass configuration for local storage devices
+    // +Optional
+    Storage Storage `json:"storage,omitempty"`
+  }
+  type Storage struct {
+    // DeviceClasses are a rules that assign local storage devices to volumegroups that are used for creating lvm based PVs
+    // +Optional
+    DeviceClasses []DeviceClass `json:"deviceClasses,omitempty"`
+  }
+  
+  type DeviceClass struct {
+    // Name of the class, the VG and possibly the storageclass.
+    // Validations to confirm that this field can be used as metadata.name field in storageclass
+    // ref: https://github.com/kubernetes/apimachinery/blob/de7147/pkg/util/validation/validation.go#L209
+    // +kubebuilder:validation:MaxLength=245
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
+    Name string `json:"name,omitempty"`
+    
+    // DeviceSelector is a set of rules that should match for a device to be included in the LVMCluster
+    // +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    
+    // NodeSelector chooses nodes on which to create the deviceclass
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
```

Comment:
Will node selection work properly with the Subprovisioner CSI plugin? My understanding is that TopoLVM has implemented support to make pod scheduling work when volumes are available only on specific nodes. Subprovisioner has no scheduler awareness and just runs a CSI Node Plugin `DaemonSet` on all nodes, expecting any node to be able to run pods that access any volume.

I see three possibilities:
1. `NodeSelector` already works with Subprovisioner, somehow. There's no work to do.
2. `NodeSelector` doesn't work with Subprovisioner and must be nil for now.
3. `NodeSelector` doesn't work with Subprovisioner yet but we figure out how to make Subprovisioner support it, so this syntax is allowed with Subprovisioner.

Reply:
The way to implement this is to implement a Node accessibility constraint in the CSI driver and to start supporting node topologies via CSI and at the same time use the node selector to only deploy on the nodes we want.

Reply:
It's possible that users would want the ability to use a subset of nodes, so I guess Subprovisioner should implement this.